flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializer.java

```diff
@@ -161,7 +162,93 @@
+	 *     has been reconfigured to be compatible, to continue reading old data, and that the
+	 *     serialization schema remains the same. No migration needs to be performed.</li>
+	 *
+	 *     <li>{@link MigrationStrategy#migrateWithFallbackDeserializer(TypeSerializer)}: this signals Flink that
+	 *     migration needs to be performed, because this serializer is not compatible, or cannot be reconfigured to be
+	 *     compatible, for old data. Furthermore, in the case that the preceding serializer cannot be found or
+	 *     restored to read the old data, the provided fallback deserializer can be used.</li>
+	 *
+	 *     <li>{@link MigrationStrategy#migrate()}: this signals Flink that migration needs to be performed, because
+	 *     this serializer is not compatible, or cannot be reconfigured to be compatible, for old data.</li>
+	 * </ul>
+	 *
+	 * <p>This method is guaranteed to only be invoked if the preceding serializer's configuration snapshot is not the
+	 * singleton {@link ForwardCompatibleSerializationFormatConfig#INSTANCE} configuration. In such cases, Flink always
+	 * assume that the migration strategy is {@link MigrationStrategy#migrate()}.
+	 *
+	 * @see MigrationStrategy
+	 *
+	 * @param configSnapshot configuration snapshot of a preceding serializer for the same managed state
+	 *
+	 * @return the result of the reconfiguration.
+	 */
+	protected abstract MigrationStrategy<T> getMigrationStrategy(TypeSerializerConfigSnapshot configSnapshot);
+
+	/**
+	 * Get the migration strategy to use this serializer based on the configuration snapshot of a preceding
+	 * serializer that was registered for serialization of the same managed state (if any - this method is only
+	 * relevant if this serializer is registered for serialization of managed state).
+	 *
+	 * <p>This method is not part of the public user-facing API, and cannot be overriden. External operations
+	 * providing a configuration snapshot of preceding serializer can only do so through this method.
+	 *
+	 * <p>This method always assumes that the migration strategy is {@link MigrationStrategy#noMigration()} if
+	 * the provided configuration snapshot is the singleton {@link ForwardCompatibleSerializationFormatConfig#INSTANCE}.
+	 * Otherwise, the configuration snapshot is provided to the actual
+	 * {@link #getMigrationStrategy(TypeSerializerConfigSnapshot)} (TypeSerializerConfigSnapshot)} implementation.
+	 *
+	 * @param configSnapshot configuration snapshot of a preceding serializer for the same managed state
+	 *
+	 * @return the result of the reconfiguration.
+	 */
+	@Internal
+	public final MigrationStrategy<T> getMigrationStrategyFor(TypeSerializerConfigSnapshot configSnapshot) {
+		// reference equality is viable here, because the forward compatible
+		// marker config will always be explicitly restored with the singleton instance
+		if (configSnapshot != ForwardCompatibleSerializationFormatConfig.INSTANCE) {
```

Comment:
I see the intention, but I think it is not a good idea to have this `final`default implementation around a `ForwardCompatibleSerializationFormatConfig`.

Imagine a case where the old version of a serializer is `ForwardCompatible` and the new serializer which replaces it is not. Then this will always bypass the check, even if it shouldn't.